tests/net/lib/dns_cache/src/main.c

```diff
@@ -0,0 +1,135 @@
+	struct dns_addrinfo info_read[TEST_DNS_CACHE_SIZE] = {0};
+	const char *query = "example.com";
+
+	for (size_t i = 0; i < TEST_DNS_CACHE_SIZE; i++) {
+		zassert_ok(dns_cache_add(&test_dns_cache, query, &info_write,
+					 TEST_DNS_CACHE_DEFAULT_TTL),
+			   "Cache entry adding should work.");
+	}
+	zassert_ok(dns_cache_flush(&test_dns_cache));
+	zassert_equal(0, dns_cache_find(&test_dns_cache, query, info_read, TEST_DNS_CACHE_SIZE));
+	zassert_equal(0, info_read[TEST_DNS_CACHE_SIZE - 1].ai_family);
+}
+
+ZTEST(net_dns_cache_test, test_fill_cache_to_small)
+{
+	struct dns_addrinfo info_write = {.ai_family = AF_INET};
+	struct dns_addrinfo info_read[TEST_DNS_CACHE_SIZE - 1] = {0};
+	const char *query = "example.com";
+
+	for (size_t i = 0; i < TEST_DNS_CACHE_SIZE; i++) {
+		zassert_ok(dns_cache_add(&test_dns_cache, query, &info_write,
+					 TEST_DNS_CACHE_DEFAULT_TTL),
+			   "Cache entry adding should work.");
+	}
+	zassert_equal(-ENOSR,
+		      dns_cache_find(&test_dns_cache, query, info_read, TEST_DNS_CACHE_SIZE - 1));
+	zassert_equal(AF_INET, info_read[TEST_DNS_CACHE_SIZE - 2].ai_family);
+}
+
+ZTEST(net_dns_cache_test, test_closest_expiry_removed)
+{
+	struct dns_addrinfo info_write = {.ai_family = AF_INET};
+	struct dns_addrinfo info_read = {0};
+	const char *closest_expiry = "example.com";
+
+	zassert_ok(dns_cache_add(&test_dns_cache, closest_expiry, &info_write,
+				 TEST_DNS_CACHE_DEFAULT_TTL),
+		   "Cache entry adding should work.");
+	k_sleep(K_MSEC(1));
+	for (size_t i = 0; i < TEST_DNS_CACHE_SIZE; i++) {
+		zassert_ok(dns_cache_add(&test_dns_cache, "example2.com", &info_write,
+					 TEST_DNS_CACHE_DEFAULT_TTL),
+			   "Cache entry adding should work.");
+	}
+	zassert_equal(0, dns_cache_find(&test_dns_cache, closest_expiry, &info_read, 1));
+	zassert_equal(0, info_read.ai_family);
+}
+
+ZTEST(net_dns_cache_test, test_expired_entries_removed)
+{
+	struct dns_addrinfo info_write = {.ai_family = AF_INET};
+	struct dns_addrinfo info_read[3] = {0};
+	const char *query = "example.com";
+
+	zassert_ok(dns_cache_add(&test_dns_cache, query, &info_write, TEST_DNS_CACHE_DEFAULT_TTL),
+		   "Cache entry adding should work.");
+	zassert_ok(
+		dns_cache_add(&test_dns_cache, query, &info_write, TEST_DNS_CACHE_DEFAULT_TTL * 2),
+		"Cache entry adding should work.");
+	zassert_ok(
+		dns_cache_add(&test_dns_cache, query, &info_write, TEST_DNS_CACHE_DEFAULT_TTL * 3),
+		"Cache entry adding should work.");
+	zassert_equal(3, dns_cache_find(&test_dns_cache, query, info_read, 3));
+	zassert_equal(AF_INET, info_read[0].ai_family);
+	k_sleep(K_MSEC(TEST_DNS_CACHE_DEFAULT_TTL * 1000 + 1));
+	zassert_equal(2, dns_cache_find(&test_dns_cache, query, info_read, 3));
+	zassert_equal(AF_INET, info_read[0].ai_family);
+	k_sleep(K_MSEC(TEST_DNS_CACHE_DEFAULT_TTL * 1000 + 1));
+	zassert_equal(1, dns_cache_find(&test_dns_cache, query, info_read, 3));
+	zassert_equal(AF_INET, info_read[0].ai_family);
+	k_sleep(K_MSEC(1));
```

Comment:
@jukkar 

So that some time can pass between the two calls to `dns_cache_find` and the found entry count should still remain the same since not enough time has passed.

I can remove it if you want in this particular case i don't think it makes much of a difference its only semantically more correct in my opinion.

Reply:
These sleeps seem to wait for expiration, so that looks like something we cannot avoid. Could we make the sleeps smaller, like <1sec?

Reply:
The problem is, that the TTL in DNS is expressed in seconds. We could internally store it in milliseconds and then adjust the test accordingly. Do you think this would be a good idea? It is an antipattern to adjust code to better fit the tests but there would be a significant test time decrease.

Alternatively is there a way to speed up time somehow?

Reply:
We could limit the testing scope to `native_sim` platform only, which does not execute tests in real time (with `integration_platforms` option). Thoughts?

Reply:
Sounds good to me 👍 

@jukkar @rlubos  I have adjusted the `testcase.yaml` accordingly and the checks are green.